### ISSUE TITLE
Can't add vm to an existing availability set through command line.

### ISSUE THREAD

**Issue by greenmang0** [Azure/azure-xplat-cli#1937]
The `azure vm` command does not provide any interface to add `vm` to an existing `availability set`.

**Comment by anuchandy** [comment]
these param sholud help:
--availabilty-set argument (in asm) or --avail-set in argument (in arm)

**Comment by greenmang0** [comment]
I am using `arm` mode.  I checked `azure vm -h`, couldn't find any option to add an existing `vm` to an existing `availability set`. The `availset` command which is available in `arm` mode, is just for managing `availability sets`.

**Comment by anuchandy** [comment]
@greenmang0 , i mean --avail-set option in vm create command.

![image](https://cloud.githubusercontent.com/assets/1471612/8807151/7b7fddfe-2f8f-11e5-8775-fc815d673cdf.png)

**Comment by greenmang0** [comment]
@anuchandy,  Yes, I get that. When I create a new `vm` I can add it to the existing availability set using above option, but can I  add **running** `vm` to the existing availability set? I tried it with `vm create`, but it complains that `vm` already exists.

**Comment by anuchandy** [comment]
@greenmang0 . yes right now we have only command to associate vm with an availability set during creation time. We don't have command to set it on already created vm. We will investigate and enable command for this, let us keep this issue open.

**Comment by greenmang0** [comment]
Okay

**Comment by jeffpatton1971** [comment]
This thread is a little old, but it appears that there is no way currently to add a vm to an existing availabilityset. I have tried doing this in PowerShell with no success, the error returned from the server was changing property 'availabilityset.id' is not allowed
Basically what I did was grab the availabilitysetreference from a vm that was in the proper set, and then setting the vm.availabilitysetreference equal to that value. PowerShell get-member said that property is get/set able. then I ran update-azurevm and that's what returned the error.
If anyone else has had any success it would greatly be appreciated.

**Comment by jeffpatton1971** [comment]
le sigh..
# Jeff Patton@C02PX3KSG8WL | 16:51:44 | 10-05-2015 | F:\Projects\Lowes\ops\Azure\PowerShell $  [master ≡]
$VM.AvailabilitySetReference |gm


   TypeName: Microsoft.Azure.Management.Compute.Models.AvailabilitySetReference

Name         MemberType Definition
----         ---------- ----------
Equals       Method     bool Equals(System.Object obj)
GetHashCode  Method     int GetHashCode()
GetType      Method     type GetType()
ToString     Method     string ToString()
ReferenceUri Property   string ReferenceUri {get;set;}


# Jeff Patton@C02PX3KSG8WL | 16:51:47 | 10-05-2015 | F:\Projects\Lowes\ops\Azure\PowerShell $  [master ≡]
$VM.AvailabilitySetReference.ReferenceUri = $AVset.Id
# Jeff Patton@C02PX3KSG8WL | 16:52:04 | 10-05-2015 | F:\Projects\Lowes\ops\Azure\PowerShell $  [master ≡]
Update-AzureVM -VM $VM -ResourceGroupName $VM.ResourceGroupName
Update-AzureVM : PropertyChangeNotAllowed: Changing property 'availabilitySet.id' is not allowed.
OperationID : 'c3b6533d-cdc3-4206-9987-0838ea36dd0d'
At line:1 char:1
+ Update-AzureVM -VM $VM -ResourceGroupName $VM.ResourceGroupName
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : CloseError: (:) [Update-AzureVM], ComputeCloudException
    + FullyQualifiedErrorId : Microsoft.Azure.Commands.Compute.UpdateAzureVMCommand

**Comment by bitsofinfo** [comment]
Is this available yet? How can I add an existing VM to an availability set via the CLI in ARM mode? Annoying

**Comment by damienmci** [comment]
I too am stuck with an existing VM needing to be moved into an existing Availability Set via ARM PowerShell.
Currently using 0.9.8 PowerShell in ARM mode.
Any update on this?

**Comment by sepiroth887** [comment]
bump. Any traction on this? Or even with PS?

**Comment by yugangw-msft** [comment]
@sauryadas , @huangpf , please comment

**Comment by huangpf** [comment]
This feature is TBD on the service side, and I believe there is an item for it in the latest planning. 

@sauryadas Could you please help check it with Mahesh, and share the timeline, if feasible? Thanks.

**Comment by jeffpatton1971** [comment]
Thanks @huangpf a timeline on when this may become available would be amazing.

**Comment by FdezRomero** [comment]
@sauryadas @huangpf Without being able to re-assign an existing VM to an Availability Set, nor via the Azure Portal nor the CLI, how are we supposed to achieve the Azure SLA? Recreating the VMs?

**Comment by sauryadas** [comment]
Mahesh @mahthi will provide an update when he is back

**Comment by chrisharrison** [comment]
So basically there is no way for me to add a VM to a load balancer without shutting down and completely re-creating the VM?

This is a single click in AWS.

**Comment by greenmang0** [comment]
@chrisharrison, 

You are comparing apples and oranges :)

**Comment by despairblue** [comment]
@greenmang0 You mean azure is not comparable to AWS? Would you mind to elaborate that a little more? I think I can't follow you. :)

**Comment by bitsofinfo** [comment]
+1 @greenmang0

**Comment by sijday** [comment]
FFS

**Comment by RobHudson72** [comment]
Pinging on this issue... it looks like I will have to recreate the VMs from scratch?

Ouch... ouch ouch ouch

**Comment by amarzavery** [comment]
@huangpf - do we have commands that support this?

Thanks,
Amar Zavery

On Feb 16, 2016, at 5:58 AM, "RobHudson72" <notifications@github.com<mailto:notifications@github.com>> wrote:


Pinging on this issue... it looks like I will have to recreate the VMs from scratch?

Ouch... ouch ouch ouch

-

**Comment by jbondc** [comment]
I just migrated a 'classic vm' (database) which is now in production to the new ARM model.

Created the availability set but can't add 'db1' and 'db2' to the availability set. Hmm.
If my memory is correct, this worked fine in 'classic vm mode'.

**Comment by huangpf** [comment]
@mahthi @sauryadas Please help confirm, but I think it's an ```IAAS-TBD``` feature on both the service and client sides.

**Comment by PSDhao** [comment]
bump, having the same issue.

**Comment by gdhagger** [comment]
I can't even add a NEW vm to an existing availability set using the CLI - is this working for anyone?

**Comment by SamirFarhat** [comment]
AFAIK, this feature may be addressed by the end of this year. It's a big challenge for the MS team to allow such operation. Changing the availability Set requires a review of the VM mobility architecture on Azure. Fore example, adding a VM in an Availability Set already containing a VM means putting it to different default domain. Becasue VM mobilty is a matter on Azure (No Live Migration), it's not an easy operation.
I have written a Powershell script which let you change the AS of an ARM VM by recreating it. Give it a try and enjoy :
https://buildwindows.wordpress.com/2016/02/25/add-or-change-an-arm-virtual-machines-availability-set/

**Comment by bitsofinfo** [comment]
sigh...

**Comment by jotunskij** [comment]
I'm having the same problem as @gdhagger. Using the CLI in arm-mode I can't seem to add a VM (upon creation) to an existing availset. I'm using the --availset-name flag for vm create, but when I view the availset in the portal there are no VMs listed there. This has a huge impact as when I try to add another VM to the same load balancer I get the following error:
"Virtual Machine /subscriptions/.../resourceGroups/.../providers/Microsoft.Compute/virtualMachines/[second created vm] is using different Availability Set than other Virtual Machines connected to the Load Balancer(s) [lb name]."

Anyone else having this issue?

**Comment by GoodbyeWorld** [comment]
I can't believe this.  I just built a new prod environment - dozens of VMs, days of work.

I go to configure my load-balancing for my web tier, and I'm told I need to have an availability set first.  OK, created an Availability set.  Great.  Just add my VMs to that and off I go, right?  NOPE.  I can't assign them to an "Availability set" post-build.  I have to redo all this work?  This is ridiculously disappointing.

If you don't have the technology to make this work, MS should at least include a warning while building VMs, warning users that once the VMs are built, they cannot be reassigned, and cannot be load-balanced without Availability sets.

**Comment by RobHudson72** [comment]
@GoodbyeWorld I agree, it is painful.  A warning at VM creation time is badly needed.  Fortunately, there are scripts available that do a pretty good job of tearing down a VM and rebuilding it attached to the Availability Set.  That's how I finally managed this.

**Comment by mahthi** [comment]
Really sorry to hear about the trouble you are going through @GoodbyeWorld<https://github.com/GoodbyeWorld>. This is definitely something that is high priority on our backlog. How can we help you in the meantime? Do you need CLI scripts that helps you reconstruct this environment?

**Comment by yan7109** [comment]
+1 @GoodbyeWorld

**Comment by chrisharrison** [comment]
+1 @GoodbyeWorld <https://github.com/GoodbyeWorld>

**Comment by despairblue** [comment]
@yan7109 @chrisharrison [github added a reaction feature](https://github.com/blog/2119-add-reactions-to-pull-requests-issues-and-comments), which serves the purpose of voting way better than commenting :smiley:

**Comment by chrisharrison** [comment]
Not if you're replying via email.

**Comment by GoodbyeWorld** [comment]
@mahthi thanks for your offer, but I rebuilt the web tier of the environment and redeployed my software.  It was frustrating, as an Azure customer, that this wasn't better thought-out by Microsoft.

Anyhow, while it's too late for me, I'd expect tools or scripts that automate the procedure and allow the user to avoid rebuilding their VMs would be very welcome by the community.

Thanks

**Comment by chrisharrison** [comment]
Dear Microsoft,

Don't waste your time on tools and scripts. Just focus on removing this
silly concept of "availability sets" and make it so that you can add any VM
to any load balancer. Via the GUI. In 2 or 3 steps. Like how AWS has done
it for years.

**Comment by neil4tell** [comment]
+1 @chrisharrison 
...and while you're at it, please add a "clone" button as well so I can make 5 more VMs just like the one I setup and put them all in the same resource group, availability set, private network, and public load balancer (without creating a complicated script). This seems like a fairly common task.

**Comment by sippylabs** [comment]
Why is this information so hard to find after _MONTHS_ of people asking for a response?

**Comment by diditho** [comment]
Oh boy, I have to rebuild all of my VMs... :-( please make this a priority

**Comment by cblaze22** [comment]
How in the world is this not an option?  This is a basic command we would need.

**Comment by amarzavery** [comment]
@mahthi @singhkay @alpon @sauryadas @huangpf - The list of customers asking this is growing? Are there any plans to fix this? Can customers get some kind of a timeline for this fix?

**Comment by mahthi** [comment]
At this time, I don’t have a locked down date to give out. This is currently targeted for H2CY2016.

**Comment by fangjinmin** [comment]
It is so terriable to confirm that can't add created vms to the Availability Set. 
recreate vms or give up Azure, I have to make a choice.

**Comment by VireshMathapati13** [comment]
You can 
there's an alternative way to do this
How to move ARM Virtual Machine into Availability Set? Follow the following link

http://blog.e-zest.net/how-to-move-arm-virtual-machine-into-availability-set

**Comment by SamirFarhat** [comment]
Viresh,
Thanks for the script. But this script will generalize the Virtual Machine image and only applicable for Windows VMs. The target VM will be different from the new one.

The following script permits moving VMs between AS

https://buildwindows.wordpress.com/2016/02/25/add-or-change-an-arm-virtual-machines-availability-set/

**Comment by VireshMathapati13** [comment]
yeah Samir.
The solution posted by me is an alternative solution.
and thanks for your solutions.

**Comment by jeffpatton1971** [comment]
Ya, @SamirFarhat has the best solution for this until something from Microsoft comes along. Its silly that we have to come up with scripts like this to create functionality that should have been there already. The only downside to ANY of these solutions is that the vm is powered off.

Although it is cool that we were able to come up with a handful of solutions, way to go guys!

**Comment by PirateBread** [comment]
I can't believe this has not been resolved yet.

Still can't move VM's into Availability sets.
Still can't load balance VM's which aren't in availability sets.

This is so backward I can't believe who thought this was a good idea.

**Comment by SamirFarhat** [comment]
Hi,
Indeed, it's a real limitation to not be able control this. But it's a hard work behind, that MS are doing to enable this.
Do not forget that it's quiet logical to have VMs within availability Set to be part of a load balancer. Load Balancing is for resource sharing and high availability. HA is not guaranteed without AS

**Comment by PirateBread** [comment]
Sure but there are plenty of times where you just want to spread the load and aren't building a fully HA setup. Like right now..... I need to spread the load for performance reasons. I had to build a haproxy VM to move traffic.

I'm a big fan of Azure but honestly basic things like this are well overdue. MS need to stop rolling out esoteric preview features and instead concentrate on the fundamentals.

Why is DNS still in preview? How long has Amazon had a stable DNS product? Like 5 years now? How is anyone meant to built out a scalable cloud service without a stable and reliable cloud DNS service?

Anyway don't get me started.... I'll just shut up and hope Microsoft eventually pull their finger out.

**Comment by FdezRomero** [comment]
@PirateBread I agree, word by word :+1:

**Comment by Findeton** [comment]
This is disappointing and frustrating.

**Comment by KpaBap** [comment]
Why is this still even open? As it stands if you miss a single small step while creating a VM you have to redo all your work and re-deploy simply to add a VM to an availability set. Add another frustrated Azure user to the list.

**Comment by singhkays** [comment]
@KpaBap Sorry for the frustration this is causing you. I've raised this customer frustration with the compute dev team previously as well. It came up for discussion for CY16H2 planning but was deferred for other high pri compute features that were launched this year and will launch early next year. Planning for CY17H1 begins soon and I'll bring this up again.

**Comment by connor-mac** [comment]
Having the same issue, would like to see a change. Please notify if anything comes up in the near future.. Extremely frustrating.

**Comment by NoureenS** [comment]
Does this help: https://docs.microsoft.com/en-us/azure/virtual-machines/virtual-machines-availability-set-supportability?

**Comment by NoureenS** [comment]
Also saw this in another forum but have not tested:


down vote

accepted
 

So... just posting this as an answer, to properly close the loop based on the comments under the question:

When setting up a Virtual Machine, you can choose which Cloud Service to place the Virtual Machine in. The Cloud Service is essentially a container which gets assigned a specific IP address, gets a cloudapp.net name (e.g. myservice.cloudapp.net), and gets assigned to a region (or affinity group, which is region-specific).

Availability Sets are specific to a given Cloud Service. You may place any of your Cloud Service's VMs in the same Availability Set (or even have multiple Availability Sets, with groups of VMs assigned to specific Availability Sets). However: An Availability Set does not span across Cloud Services.
 https://stackoverflow.com/questions/23321063/unable-to-add-a-vm-to-an-availability-set-in-azure
